cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java

```diff
@@ -0,0 +1,449 @@
+ *      private final AuthorizationEnforcer authorizationEnforcer;
+ *      private final AuthenticationContext authenticationContext;
+ *
+ *      public ValidAuthEnforceAnnotations() {
+ *        authenticationContext = new AuthenticationTestContext();
+ *        authorizationEnforcer = new ExceptionAuthorizationEnforcer();
+ *      }
+ *
+ *      public void testSingleAction(@Name("namespaceId") NamespaceId namespaceId) throws Exception {
+ *        AuthEnforceAnnotationEnforcer.enforce(this.authorizationEnforcer, namespaceId, authenticationContext,
+ *                                              Set<Action.Admin>);
+ *        System.out.println("Hello");
+ *      }
+ *     }
+ *   </pre>
+ * </p>
+ */
+public class AuthEnforceRewriter implements ClassRewriter {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AuthEnforceRewriter.class);
+
+  private static final Type AUTHORIZATION_ENFORCER_TYPE = Type.getType(AuthorizationEnforcer.class);
+  private static final Type AUTHENTICATION_CONTEXT_TYPE = Type.getType(AuthenticationContext.class);
+  private static final Type ACTION_TYPE = Type.getType(Action.class);
+  private static final Type AUTH_ENFORCE_ANNOTATION_ENFORCER_TYPE = Type.getType(AuthEnforceAnnotationEnforcer.class);
+  private Map<Method, AnnotationDetails> methodAnnotations;
+
+  @Override
+  public byte[] rewriteClass(String className, InputStream input) throws IOException {
+    byte[] classBytes = ByteStreams.toByteArray(input);
+    methodAnnotations = new HashMap<>(); // re-initialize for this new class.
+
+    // First pass: Check the class to have a method with AuthEnforce annotation if found store the annotation details
+    // and parameters for the method for second pass in which class rewrite will be performed.
+    ClassReader cr = new ClassReader(classBytes);
+
+    // 0 because we don't want the class writer to compute frames or maxs. Since
+    // in first pass we just pass the classes for annotation. // Note: There is no need to compute frames in this
+    // pass since we are not generating any byte code and for classes which does not have AuthEnforce annotation we
+    // just return original byte[] from the passed input stream. Trying to compute frames require loading all the
+    // dependent classes for a class and this might lead to class loading problems if some dependent classes are not
+    // available. For example HbaseTableUtil use Hbase classes and in in-memory and standalone cdap hbase classes
+    // are not present.
+    ClassWriter cw = new ClassWriter(0);
+
+    // SKIP_CODE to make the first pass faster since in the first pass we just want to process annotations to check
+    // if the class has any method with AuthEnforce annotation. If such method is found we also store the parameters
+    // which has the named annotations as specified in the entities field of the AuthEnforce.
+    cr.accept(new AuthEnforceAnnotationVisitor(className, cw), ClassReader.SKIP_CODE);
```

Comment:
Should skip everything that can be skipped: `SKIP_CODE | SKIP_DEBUG | SKIP_FRAMES`